ENH: improve dlpack protocol detection

gigony opened the pull request:
Currently, dlpack protocol detection is done by checking the existence of `__dlpack__` method on the object's type.

This is not enough to detect the dlpack protocol support, because the method can be defined on the object itself, not on the object's type.

This patch improves the detection by checking the existence of `__dlpack__` method on the object itself.

This patch allows the following code:

```python
    def test_from_dlpack_dynamic_method(self):
        class TensorWrapper:
            def __init__(self, tensor):
                self.tensor = tensor
                if not hasattr(self, '__dlpack__'):
                    self.__dlpack__ = tensor.__dlpack__
        x = np.arange(5)
        y = TensorWrapper(x)
        z = np.from_dlpack(y)
        assert_array_equal(x, z)
```

FYI, CuPy checks `__dlpack__` method on the object itself, not on the type of the object so it doesn't have a problem with the code like above.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

gigony committed:
ENH: improve dlpack protocol detection

Currently, dlpack protocol detection is done by checking the existence
of `__dlpack__` method on the object's type. This is not enough to
detect the dlpack protocol support, because the method can be defined on
the object itself, not on the object's type.
This patch improves the detection by checking the existence of
`__dlpack__` method on the object itself.

Signed-off-by: Gigon Bae <gigony@gmail.com>

charris commented:
This could use a release note, see `doc/release/upcoming_changes` to see how that is done.

Might also be worth looking at the documentation in `doc/source/user/basics.interoperability.rst` to see if it could use updating.

@hameerabbasi Thoughts?

hameerabbasi commented:
I'm also having trouble thinking of a non-edge use-case for this. I think this it a bug to allow this on the CuPy side. Dunder methods are usually defined on the type.

Even if we call e.g. `__array_function__` on the object, it was too late to change it at that point, and it should really have been done on the type. I would not go down this road with `__dlpack__`. It opens the road to value-based semantics, rather than type-based semantics.

hameerabbasi commented:
While not natural, it can be made a `classmethod`, which is the same on the type and the array.

leofang commented:
1. I don't think CuPy (and a dozen of other libraries) did the wrong thing. We did say in the [protocol](https://dmlc.github.io/dlpack/latest/python_spec.html#syntax-for-data-interchange-with-dlpack) that we'd check if the _object_ has the dunder method. So, I think what's odd (and I missed this during the review) was NumPy only accepted `o->ob_type` instead of just `o` as @gigony showed in the above patch (which I like a lot better than the present PR TBH, I had trouble understanding the argument).
2. However, the example in the PR description (and also added as a test) `test_from_dlpack_dynamic_method` makes me cringe. It is as if we encourage using DLPack to exchange an object not owning the underlying buffer to be consumed by another side. Can we please use a better example/test?

gigony commented:
Hi @charris ,
I have updated `doc/release/upcoming_changes/23360.improvement.rst` file here: https://github.com/numpy/numpy/pull/23360/files#diff-32f4012297c7f9685801c8f055cc7dd1b26e2625ce482be6a8c8f0a6cfb68833

Since the existing [`basics.interoperability.rst` document](https://github.com/numpy/numpy/blob/main/doc/source/user/basics.interoperability.rst#the-dlpack-protocol) already mentions that "on the array object", not "on the array object's type", I don't think the change is needed there.
> 2. **`__dlpack__(self, stream=None)` and `__dlpack_device__` methods on the array object**, which will be called from within from_dlpack, to query what device the array is on (may be needed to pass in the correct stream, e.g. in the case of multiple GPUs) and to access the data.

Thanks @leofang for the feedback.

I have updated the current PR to simplify the change (using `obj` instead of `obj->ob_type`) and updated example/test to be more realistic:
```c
PyObject *capsule = PyObject_CallMethod((PyObject *)obj, "__dlpack__", NULL);
```
```python
    def test_from_dlpack_dynamic_method(self):
        class TensorWrapper:
            def __init__(self, tensor):
                self.tensor = tensor
                if not hasattr(self, '__dlpack__'):
                    self.__dlpack__ = tensor.__dlpack__
        x = np.arange(5)
        y = TensorWrapper(x)
        z = np.from_dlpack(y)
        assert_array_equal(x, z)
```

leofang commented:
Thanks for catching / working on this issue, Gigon, LGTM pending tests, but NumPy experts should make the final call 🙂

hameerabbasi commented:
I won't block this if the standard specifies it and 20+ libraries already do it -- it's already too late at that point to change. 😉

seberg commented:
A bit in the same boat as Hameer here, because it seems to me that this mainly forces a cleaner implementation using a property/method; and that protects us from cringing.  I have reached out to propose a downstream fix (also means it works on old NumPy versions).

If it isn't as simple as I am hoping, I am mildly in favor of just doing this (and will merge).  If the downstream fix is straight forward, I may need another gentle nudge to press the green button ;).

seberg commented:
(I admit, this may be a false sense of "clean" on my part, since Python dunders have a C-slot that needs filling.  But we don't have that C-counterpart here.)

leofang commented:
Hi @seberg just wanna follow up: what's the downstream fix that you referred to, and how is it going?

seberg commented:
Gigon had already changed things to use a property, not dynamically adding an attribute.  I think that is a bit nicer anyway, but doesn't mean we can't change it here, though.

leofang commented:
Friendly ping here, should we merge or close this PR?

seberg commented:
Thanks @gigony, I slightly prefer to keep the current pattern, so going to close.  I have to admit: that may just be me being misguided about trying to mirror C-operators here (although I have found myself wishing we did it in similar methods, I am not sure the reasons apply here).

In other words: *I am very happy to reopen and merge if anyone voices a clear preference for the other direction.*
